tutel/impls/moe_layer.py

```diff
@@ -134,13 +134,22 @@ def apply_on_expert_fn(self, input, expert_fn, group, sharded_count):
 
         dispatched_input = self._fdr.encode(input)
         dispatched_input = dispatched_input.repeat(sharded_count, 1)
-        dispatched_input = AllToAll.apply(group, dispatched_input)
         dispatched_input = dispatched_input.reshape(world_size, -1, capacity, M)
 
-        expert_output = expert_fn(dispatched_input)
-        expert_output = expert_output.to(input.dtype)
+        num_split = 2
+        split_dim = 2
+        AllToAllStatus.init(group, num_split, split_dim, dispatched_input)
+
+        # Implicit x.contiguous() in CurrentStreamRelease.forward() and CurrentStreamAcquire.backward()
+        dispatched_input_ready = CurrentStreamRelease.apply(dispatched_input, 0)
+        dispatched_input_scattered_after_a2a = AllToAllScatterAsync.apply(dispatched_input_ready)
+
+        expert_output_scattered = [
+            CurrentStreamRelease.apply(expert_fn(CurrentStreamAcquire.apply(x, i)).to(input.dtype), i)
+            for i, x in enumerate(dispatched_input_scattered_after_a2a)]
+        expert_output_gathered_after_a2a = AllToAllGatherAsync.apply(*expert_output_scattered)
+        expert_output = CurrentStreamAcquire.apply(expert_output_gathered_after_a2a, 0)
 
```

Comment:
Is it possible to set an option to enable this instead of directly replacing the original approach?

Reply:
Sure, will make this an option in formal PR later.